File: src/game/entity/enemies/boss/clown/Hat.java
package game.entity.enemies.boss.clown;

import game.entity.Player;
import game.entity.enemies.npcs.Orb;
import game.entity.models.Entity;
import game.entity.models.Coordinates;
import game.entity.models.Direction;
import game.entity.models.EnhancedDirection;
import game.utils.Paths;
import game.values.DrawPriority;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class Hat extends Orb {
    @Override
    protected String getBasePath() {
        return Paths.getEnemiesFolder() + "/clown/hat";
    }

    @Override
    public String[] getCharacterOrientedImages() {
        return new String[]{
                getBasePath() + "1.png",
                getBasePath() + "2.png",
                getBasePath() + "3.png",
                getBasePath() + "4.png",
                getBasePath() + "5.png",
                getBasePath() + "6.png",
                getBasePath() + "7.png",
                getBasePath() + "8.png",
                getBasePath() + "9.png",
                getBasePath() + "10.png",
        };
    }

    public Hat(Coordinates coordinates, EnhancedDirection enhancedDirection) {
        super(coordinates, enhancedDirection);
        setMaxHp(300);
        setHp(getMaxHp());
    }

    @Override
    public int getSize() {
        return SIZE * 3;
    }

    @Override
    protected void doInteract(Entity e) {
        if(e == null) return;

        if(e instanceof Clown) {
            ((Clown) e).setHasHat(true);
            despawn();
        } else {
            attack(e);
        }
    }

    @Override
    protected boolean useOnlyBaseIcons() {
        return true;
    }

    protected void updateDirection() {
        if(!canMove || !isAlive) return;

        if (enhancedDirection == null) {
            // When hitting a wall, bounce and change direction;
            if(!moveOrInteract(direction))
                direction = direction.opposite();

            updateLastDirection(direction);

            return;
        }

        for (Direction d : enhancedDirection.toDirection()) {
            currDirection = d;
            if (!moveOrInteract(d)) {
                enhancedDirection = enhancedDirection.opposite(d);
            }

            updateLastDirection(currDirection);
        }
    }

    @Override
    public Set<Class<? extends Entity>> getInteractionsEntities(){
        return new HashSet<>(Arrays.asList(Player.class, Clown.class));
    }

    @Override
    public Set<Class<? extends Entity>> getObstacles(){
        return new HashSet<>();
    }

    @Override
    public void doUpdate(boolean gameState) {
        if(gameState) updateDirection();
    }

    @Override
    public DrawPriority getDrawPriority() {
        return DrawPriority.DRAW_PRIORITY_3;
    }
}

File: src/game/entity/enemies/boss/clown/Hat.kt
package game.entity.enemies.boss.clown

import game.entity.EntityTypes
import game.entity.player.Player
import game.entity.enemies.npcs.Orb
import game.entity.models.Coordinates
import game.entity.models.EnhancedDirection
import game.entity.models.Entity
import game.utils.Paths
import game.values.DrawPriority
import java.util.*

open class Hat : Orb {
    constructor(coordinates: Coordinates?, enhancedDirection: EnhancedDirection?) : super(coordinates, enhancedDirection) {
        initHat()
    }

    constructor() : this(null, null)

    constructor(id: Long) : super(id)

    private fun initHat() {
        maxHp = 300
    }

    override val entitiesAssetsPath: String get() ="${Paths.enemiesFolder}/clown/hat"
    override val drawPriority: DrawPriority
        get() = DrawPriority.DRAW_PRIORITY_3

    override val type: EntityTypes
        get() = EntityTypes.Hat

    override fun getInteractionsEntities(): Set<Class<out Entity>> = hashSetOf(Player::class.java, Clown::class.java)

    override fun getObstacles(): Set<Class<out Entity>> = HashSet()

    override fun getCharacterOrientedImages(): Array<String> = Array(10) { index ->
        "${entitiesAssetsPath}${index + 1}.png"
    }

    override val size: Int
        get() = SIZE * 3

    override fun doInteract(e: Entity?) {
        if (e == null) return

        if (e is Clown) {
            e.hasHat = true
            despawnAndNotify()
        } else {
            attack(e)
        }
    }

    override fun useOnlyBaseIcons(): Boolean = true

    private fun updateDirection() {
        if (!canMove || !isAlive)
            return

        if (enhancedDirection == null) {
            // When hitting a wall, bounce and change direction;
            if (!moveOrInteract(direction))
                direction = direction!!.opposite()

            updateLastDirection(direction)
            return
        }

        for (d in enhancedDirection!!.toDirection()) {
            currDirection = d
            if (!moveOrInteract(d)) {
                enhancedDirection = enhancedDirection!!.opposite(d)
            }
            updateLastDirection(currDirection)
        }
    }

    override fun doUpdate(gameState: Boolean) {
        if (gameState) updateDirection()
    }
}
